features/admin.rules.v1/providers/rules-provider.tsx

```diff
@@ -0,0 +1,408 @@
+/**
+ * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SelectChangeEvent } from "@mui/material";
+import React, {
+    ReactNode,
+    createContext,
+    useContext,
+    useEffect,
+    useState } from "react";
+import { v4 as uuidv4 } from "uuid";
+import {
+    ConditionTypes,
+    ExpressionInterface,
+    RuleComponentMetaDataInterface,
+    RuleConditionsInterface,
+    RuleExecutionMetaDataInterface,
+    RuleInterface
+} from "../models/rules";
+
+/**
+ * Interface for the RulesContext.
+ */
+interface RulesContextInterface {
+    /**
+     * Rules instance.
+     */
+    rulesInstance: RuleInterface[];
+
+    /**
+     * Conditions meta.
+     */
+    conditionsMeta: any;
+
+    /**
+     * Condition expressions meta.
+     */
+    conditionExpressionsMeta: RuleComponentMetaDataInterface;
+
+    /**
+     * Method to add a new rule.
+     */
+    addNewRule: () => void;
+
+    /**
+     * Method to remove a rule.
+     */
+    removeRule: (id: string) => void;
+
+    /**
+     * Method to add a new rule condition.
+     */
+    addNewRuleCondition: (ruleId: string, previousConditionInstanceId: string, conditionType: ConditionTypes) => void;
+
+    /**
+     * Method to remove a rule condition.
+     */
+    removeRuleCondition: (ruleId: string, conditionId: string) => void;
+
+    /**
+     * Method to update the rule execution type.
+     */
+    updateRuleExecution: (event: SelectChangeEvent, id: string) => void;
+
+    /**
+     * Method to update the rule condition expression value.
+     */
+    updateRuleConditionExpression: (
+        changedValue: string,
+        ruleId: string,
+        conidtionId: string,
+        expressionId: string,
+        fieldName: keyof ExpressionInterface) => void;
+}
+
+/**
+ * Create the context
+ */
+const RulesContext: React.Context<RulesContextInterface | undefined> =
+    createContext<RulesContextInterface | undefined>(undefined);
+
+// Refference to hold the latest context value
+const RuleContextRef: { ruleInstance: RuleInterface[] | undefined } = { ruleInstance: undefined };
+
+/**
+ * Method to get the context value
+ *
+ * @returns RuleInstanceData
+ */
+export const getRuleContextValue = () => RuleContextRef.ruleInstance;
```

Comment:
Shall we move this to a separate hook ex: `useRule()` and access the context from there?

Reply:
Yes, it's better. Will change it.